test/services/database/challenge_repository_test.dart

```diff
@@ -0,0 +1,206 @@
+import "package:cloud_firestore/cloud_firestore.dart";
+import "package:fake_cloud_firestore/fake_cloud_firestore.dart";
+import "package:flutter_test/flutter_test.dart";
+import "package:proxima/models/database/challenge/challenge_data.dart";
+import "package:proxima/models/database/challenge/challenge_firestore.dart";
+import "package:proxima/models/database/post/post_data.dart";
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/models/database/user/user_firestore.dart";
+import "package:proxima/services/database/challenge_repository_service.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+
+import "../../mocks/data/firestore_user.dart";
+import "../../mocks/data/geopoint.dart";
+import "../../mocks/data/post_data.dart";
+
+Future<List<PostData>> addPosts(
+  PostRepositoryService postRepository,
+  GeoPoint pos,
+  int n,
+) async {
+  final fakePosts = PostDataGenerator.generatePostData(n);
+  for (final post in fakePosts) {
+    await postRepository.addPost(post, pos);
+  }
+  return fakePosts;
+}
+
+void main() {
+  late FakeFirebaseFirestore firestore;
+  late PostRepositoryService postRepository;
+  late ChallengeRepositoryService challengeRepository;
+
+  setUp(() async {
+    firestore = FakeFirebaseFirestore();
+    postRepository = PostRepositoryService(firestore: firestore);
+    challengeRepository = ChallengeRepositoryService(
+      firestore: firestore,
+      postRepositoryService: postRepository,
+    );
+  });
+
+  const userPos = userPosition1;
+  final inChallengeRange = GeoPointGenerator.createOnEdgeInsidePosition(
+    userPos,
+    ChallengeRepositoryService.maxChallengeRadius,
+  );
+
+  final uid = testingUserFirestoreId;
+
+  group("ChallengeRepositoryService", () {
+    test("Get new challenges", () async {
+      final fakePosts = await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      expect(challenges.length, ChallengeRepositoryService.maxActiveChallenges);
+      for (final challenge in challenges) {
+        expect(challenge.data.isCompleted, false);
+        expect(challenge.data.expiresOn.compareTo(Timestamp.now()), isPositive);
+
+        final actualPost = await postRepository.getPost(challenge.postId);
+        expect(fakePosts.contains(actualPost.data), true);
+      }
+    });
+
+    test("Complete a challenge", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
```

Comment:
I think it is worth verifying also here that the challenge list has length 1, since you also do it later:

```suggestion
  expect(challenges.length, 1);

```